src/Orleans.Streaming.Redis/Storage/RedisDataManager.cs

```diff
@@ -8,50 +8,83 @@
 using System.Linq;
 using Orleans.Providers.Streams.Redis;
 using Orleans.Configuration;
+using Serilog.Core;
+using Orleans.Redis.Common;
+using System.Threading;
 
 namespace Orleans.Streaming.Redis.Storage
 {
-    internal class RedisDataManager
+    internal class RedisDataManager : IRedisDataManager
     {
+        private const int MINIMUM_KEY_LENGTH = 5;
+        private const int MAXIMUM_KEY_LENGTH = 256;
+
         public const int UnlimitedMessageCount = -1;
 
         public string QueueName { get; private set; }
 
         private readonly RedisStreamOptions _options;
+        private readonly IConnectionMultiplexerFactory _connectionMultiplexerFactory;
         private readonly ILogger _logger;
 
         private readonly ConcurrentQueue<RedisValue> _queue = new ConcurrentQueue<RedisValue>();
         private readonly RedisChannel _redisChannel;
 
         private static readonly IEnumerable<RedisValue> EmptyRedisValueEnumerable = Array.Empty<RedisValue>();
 
+        private bool _initialized = false;
         private IConnectionMultiplexer _connectionMultiplexer;
 
-        private RedisDataManager(RedisStreamOptions options, ILogger loggerFactory, string queueName)
+        internal ConcurrentQueue<RedisValue> TestHook_Queue => _queue;
+
+        private RedisDataManager(RedisStreamOptions options, IConnectionMultiplexerFactory connectionMultiplexerFactory, ILogger loggerFactory, string queueName)
         {
             queueName = SanitizeQueueName(queueName);
             ValidateQueueName(queueName);
 
             _options = options;
-            _logger = loggerFactory.ForContext<RedisDataManager>();
+            _connectionMultiplexerFactory = connectionMultiplexerFactory;
+            _logger = (loggerFactory ?? SilentLogger.Logger).ForContext<RedisDataManager>();
             QueueName = queueName;
 
             _redisChannel = new RedisChannel(QueueName, RedisChannel.PatternMode.Literal);
         }
 
-        public RedisDataManager(RedisStreamOptions options, ILogger loggerFactory, string queueName, string serviceId)
-            : this(options, loggerFactory, serviceId + "-" + queueName) { }
+        public RedisDataManager(RedisStreamOptions options, IConnectionMultiplexerFactory connectionMultiplexerFactory, ILogger loggerFactory, string queueName, string serviceId)
+            : this(options, connectionMultiplexerFactory, loggerFactory, serviceId + ":" + queueName) { }
 
-        public async Task InitAsync()
+        public async Task InitAsync(CancellationToken ct = default)
         {
             var startTime = DateTimeOffset.UtcNow;
 
             try
             {
-                _connectionMultiplexer = await ConnectionMultiplexer.ConnectAsync(_options.ConnectionString);
+                _connectionMultiplexer = await _connectionMultiplexerFactory.CreateAsync(_options.ConnectionString);
```

Comment:
It seems like `CreateAsync` should probably take in a `CancellationToken` since the implementations of the multiplexer factory call `ConnectAsync`.